### ISSUE TITLE
Qwen2.5-7B-Instruct-q40ef16: NaN logits (always on CPU, input-dependent on Metal); Qwen3-1.7B-q40ef16 is fine

### ISSUE THREAD

**Issue by czoli1976** [sonos/tract#2475]
`Qwen/Qwen2.5-7B-Instruct` in the CI-published **q40ef16** variant produces **NaN logits** in tract: *always* on the default CPU runtime, and *input-dependently* on Metal. The same code path with the smaller **Qwen3-1.7B-q40ef16** is perfect, so this is not a general q40ef16 problem.

Reproduces with the stock `causal_llm` example only — no custom code.

## Repro

```sh
wget https://s3.amazonaws.com/tract-ci-builds/tests/llm/current/Qwen--Qwen2.5-7B-Instruct-q40ef16/Qwen--Qwen2.5-7B-Instruct-q40ef16.nnef.tgz
wget https://huggingface.co/Qwen/Qwen2.5-7B-Instruct/resolve/main/tokenizer.json

cargo run --bin complete --release -- --force-cpu -t tokenizer.json \
    -m Qwen--Qwen2.5-7B-Instruct-q40ef16.nnef.tgz -n 8 "The capital of France is"
```

| model / runtime | result |
|---|---|
| Qwen2.5-7B-Instruct-q40ef16, **CPU** | empty / garbage — logits are all NaN on every step |
| Qwen2.5-7B-Instruct-q40ef16, **Metal** | *input-dependent*: `"The capital of France is"` → correct (`C. Paris`), but a ChatML-wrapped prompt → NaN on many steps |
| **Qwen3-1.7B-q40ef16**, CPU | ✅ correct: `"The capital of France is Paris. The capital of the United States is Washington"` |

On Metal the NaN is intermittent, which is what makes it look like garbled text rather than an outright failure — real tokens interleaved with the argmax-of-NaN token:

```
Ciao! Come posso aiutarti oggi?   ->   "! Ciao!! Come! pos! posso!!!! ai!ut!i!o!t"
```

(every `!` is token 0, i.e. `argmax` over an all-NaN logit row)

## Where it comes from

I bisected it by splitting the model into one sub-model per transformer block (28 of them) and running them in sequence, dumping `max|residual|` after each block. Prefill of a 21-token prompt, Metal:

```
block  0: max|res|      63.2  non-finite 0
block  1: max|res|      30.4  non-finite 0
...
block 25: max|res|      11.2  non-finite 0
block 26: max|res|     419.0  non-finite 0     <- 25x the others, but still far under f16 max (65504)
block 27: max|res|       0.0  non-finite 152064  <- ALL 152064 logits NaN
```

So **the residual stream is healthy end-to-end** — nothing is near the f16 ceiling — and the NaN appears only in the **last module** (block 27 + final RMSNorm + `lm_head`). It is not a gradual f16 overflow of the residual, and it is not RMSNorm: both the CPU fast path (`core/src/ops/nn/rms_norm.rs`, casts to f32 before the reduction) and both Metal kernels (`metal/src/kernels/nn/nn_ops.metal`, `rms_norm_nd3` / `rms_norm_nd2_l4`, both `static_cast<float>` before squaring) accumulate in f32.

Notably the residual jumps to ~419 at block 26 while every earlier block sits around 10-70. Qwen2.5-7B is known for large activation outliers; squaring 419 gives ~175 000, which is past the f16 max — so any op in that last block that squares/accumulates an activation in f16 rather than f32 would overflow to `inf` and then NaN. Qwen3-1.7B never gets near those magnitudes, which is consistent with it being fine.

## Suggested direction

The failure is confined to the last block + `lm_head`, on both backends, so it is a core/dtype issue rather than a backend kernel one. Two concrete things worth checking:

1. **Accumulator dtype in the last block's ops for f16 activations.** Everything that reduces/squares an activation should accumulate in f32 (as RMSNorm already does). The `q40ef16` variant is the only one with f16 activations, and it is the only one that fails.

2. **The quantized `lm_head`.** `q40ef16` is exported with `-c min_max_q4_0_with_embeddings`, so the (very wide, 3584 x 152064) head is block-quantized. It's the one op unique to the failing module.

A quick way to discriminate: run the same prompt against `Qwen--Qwen2.5-7B-Instruct-f16f16` (f16 activations, **un**quantized weights). If f16f16 is clean, the quantized head is implicated; if it also NaNs, it is the f16 activation range in that last block.

Happy to test whichever of those you'd like, or to run the per-block bisect on any other model — the tooling is easy to point at a different checkpoint.

**Comment by czoli1976** [comment]
Lost my mind on this trying to use it with Dis-Tract today

**Comment by kali** [comment]
Yeah, we've know for a while these models (and the early microsoft ones) are more susceptible to numeric instabilities than llama. TBH, my first reaction is, this is a model problem, not a bug in tract. I think it is the responsibility of the model author / cooker to generate model that do not ask the inference engine for invalid computations. Guessing the amplitude of activations by static analysis of the model is not tract work.

At some point in time, tract was decluttering RMSNorm as Sqr->Reduce, etc. and that was breaking all f16 models. That was a clear bug, we stopped doing that and introduced the reified RMSNorm so the square are not shown as wires anymore.

We also went through multiple iteration of Einsum "hints" (the "acc" parameter). That way the model has the option of forcing f32 as an Einsum accumulator on a f16 product, or other combination. I'm not 100% sure how this model is exported, and/or if tract honors this contract consistently today.

Pinging @JulienBalianSonos for his opinions and memories on this.

**Comment by JulienBalianSonos** [comment]
I pulled the current CI asset and dumped graph.nnef. The f32 hints are all present, so it's not a _stale_ export.

- all 28 tract_transformers_sdpa ops carry acc_datum_type = 'f32'
- every linear, including the quantized lm_head, is `tract_core_einsum(..., acc = 'f32')`
- the final RMSNorm reduces in f32

Baked export cmd: `-c min_max_q4_0_with_embeddings -dt f16 --reify-sdpa-operator`, transformers 4.51.3, built with `--no-verify` (so a NaN wouldn't have surfaced at export).

On T2N side this comes from a guard we've had for Qwen2.5 precisely for this NaN:

```python
if self.is_half_precision_model and self.model_infos.conf.model_type == "qwen2":
    inference_target.force_attention_inner_in_f32 = True
    inference_target.force_linear_accumulation_in_f32 = True
```

So the model is asking for f32 accumulation everywhere that matters. Since your bisect lands on the last block + lm_head, my guess is the acc = 'f32' hint isn't making it through the q4_0 matmul path (the "not sure tract honors this consistently" you mentioned), though it could be on T2N side too.

One caveat on the quantizer: it's `min_max_q4_0_with_embeddings`, i.e. naive per-block min/max with no activation-aware calibration, and the lm_head itself is quantized. **min_max** is about the most fragile choice on Qwen2.5's outliers: one outlier inflates a block's scale and wrecks the other 31 weights, worst exactly on the high-magnitude channels, which can push a logit into f16-overflow on the final cast. So min_max may be aggravating this rather than it being tract alone.

@czoli1976's f16f16 run is the clean discriminator:
- if fine and NaN only on q40ef16 → q4_0 accumulator or min_max 
- if NaN on both → an f16 spot we don't yet force to f32 (RMSNorm affine multiply, or SwiGLU gate*up), which I'd guard on our end.

**Comment by czoli1976** [comment]
It seems the ARM64 kernel the culprit and not the model. The proof is that the same model asking for the same math computes correctly on x86 CPU and on Metal, and only the arm64simd_mmm_f32_32x1_gen fused path produces NaN. A model demanding an invalid f16 computation would fail on all three; it fails on exactly one. 

Full evidence matrix — same model, same prompt, same f16 activations throughout

<img width="769" height="179" alt="Image" src="https://github.com/user-attachments/assets/0c913a85-ce35-4eb8-be0a-36912cfe8023" />

suggest to Merge #2477's guard as-is — safe correctness now - will investigate the kernel

**Comment by czoli1976** [comment]
Followed up on the kernel — root cause confirmed at the assembly level, and it's the "acc doesn't make it through the q4_0 path" @JulienBalianSonos suspected, but specifically in the **fused residual add**, not the matmul accumulation.

`arm64simd_mmm_f32_32x1_gen` has a `store(f16)` path but its `.add_unicast` never got a matching one: `.store` reads `item_size` and branches to `.store_f16` (`fcvtn`); `.add_unicast` ignores it and loads the operand as f32. When the residual `Add` fuses into the native q40 GEMV as an f16 AddUnicast, the kernel reads the f16 buffer as f32.

Instrumented the AddUnicast in `scratch.rs` — fires exactly once (the `PushSliceUp` `n=1` tail = last block):

```
operand item_size=2 (f16), row_byte_stride=2  ->  kernel TI=4 (f32), mr=32 nr=1
```

With `rsc=2`, `.add_unicast` does a 4-byte f32 load while advancing the pointer 2 bytes per element → misaligned reads of the f16 residual → `inf`/NaN. Not a real f16 overflow (residual peaks at a finite ~419).

Proof it's the kernel — all arm64 CPU, same model/prompt:

| variant | result |
|---|---|
| `main` | NaN |
| #2477 (disable the fusion) | `Paris` |
| `main` + `item_size==2`→`fcvtl` branch in `.add_unicast`, **fusion still on** | `Paris` |

The last row is the clincher: converting the f16 operand in the kernel fixes it with the fusion active. So `acc='f32'` is honored by the matmul; only the fused-unicast operand was mishandled. min_max isn't implicated — the same quantized bytes decode correctly once the kernel is fixed (and are clean on x86/Metal). Only arm64 hits this because x86's `fma_mmm_f32_32x1` isn't selected into this native-q40 fused-f16-AddUnicast tail.

**Comment by czoli1976** [comment]
@JulienBalianSonos  see https://github.com/sonos/tract/pull/2495
